src/core/transpile.jl

```diff
@@ -479,3 +484,126 @@ function transpile(::CastToPhaseShiftAndHalfRotationX, circuit::QuantumCircuit):
 
     return output_circuit
 end
+
+struct PlaceOperationsOnLine<:Transpiler end
+
+"""
+    transpile(::PlaceOperationsOnLine, circuit::QuantumCircuit)::QuantumCircuit
+
+Implementation of the `PlaceOperationsOnLine` transpiler stage 
+which adds Swap gates around multi-qubit gates so that the 
+final operator acts on adjacent qubits. The result of the input 
+and output circuit on any arbitrary state Ket is unchanged 
+(up to a global phase).
+
+# Examples
+```jldoctest
+julia> transpiler=Snowflake.PlaceOperationsOnLine();
+
+julia> circuit = QuantumCircuit(qubit_count = 6, gates=[toffoli(4,6,1)])
+Quantum Circuit Object:
+   qubit_count: 6 
+q[1]:──X──
+       |  
+q[2]:──|──
+       |  
+q[3]:──|──
+       |  
+q[4]:──*──
+       |  
+q[5]:──|──
+       |  
+q[6]:──*──
+          
+
+
+
+
+julia> transpiled_circuit=transpile(transpiler,circuit)
+Quantum Circuit Object:
+   qubit_count: 6 
+q[1]:───────────────────────────X───────────────────────────
+                                |                           
+q[2]:───────☒───────────────────*───────────────────☒───────
+            |                   |                   |       
+q[3]:──☒────☒──────────────☒────*────☒──────────────☒────☒──
+       |                   |         |                   |  
+q[4]:──☒──────────────☒────☒─────────☒────☒──────────────☒──
+                      |                   |                 
+q[5]:────────────☒────☒───────────────────☒────☒────────────
+                 |                             |            
+q[6]:────────────☒─────────────────────────────☒────────────
+                                                            
+
+
+
+julia> compare_circuits(circuit,transpiled_circuit)
+true
+
+```
+"""
+function transpile(::PlaceOperationsOnLine, circuit::QuantumCircuit)::QuantumCircuit
+
+    gates=get_circuit_gates(circuit)
+    
+    qubit_count=get_num_qubits(circuit)
+    output_circuit=QuantumCircuit(qubit_count=qubit_count)
+
+    for gate in gates
+
+        connected_qubits=sort(get_connected_qubits(gate))
+
+        if length(connected_qubits)>1
+
+            targets=get_targets(gate)
+    
+            # in controlled gates, length(target_indices)<length(connected_qubits)
```

Comment:
This transpiler should not be concerned with controlled gates.

We can have another optimizing transpiler which removes swap gates that re-order control qubits.

Reply:
This transpiler stage need to distinguish between control qubits and target qubits because when constructing the substitute with `typeof(gate)(connected_qubits_output...)` target is the last argument as per the controlled gates ctor. Both control and targets can be remapped, but the constructor arguments require correct ordering of target vs controls

Reply:
unless we convert all controlled gates constructors to `@kwdef`